Integrate `MetadataPayloadExchange` into `Shuffler`

pentschev opened the pull request:
Follow-up on #437 by switching `Shuffler` to use the new `MetadataPayloadExchange`. This helps simplify the code in `Shuffler`, and simultaneously remove the need for an "ack" message, opening also the possiblity for using an Active Message-based communicator (TBD).

pentschev committed:
Integrate `MetadataPayloadExchange` in `Shuffler`

pentschev committed:
Fix send of zero-sized payloads

madsbk reviewed (changes requested):
Overall looks great, @pentschev.

It would be good to run some benchmarks. I assume we expect the performance to remain unchanged?

pentschev committed:
Add a helper function to convert Chunks into Messages

pentschev reviewed (commented):
Thanks @madsbk , I ran `bench_shuffle` with 4 workers (I don't currently have exclusive access to an 8xH100 machine) as you suggested, and below are some results:

<details><summary>Before (run 1)</summary>

```
[PRINT:0:0] means: 0.01 sec | local throughput: 134.29 GiB/s | global throughput: 537.17 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:1:0] means: 0.01 sec | local throughput: 128.22 GiB/s | global throughput: 512.87 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:3:0] means: 0.01 sec | local throughput: 127.86 GiB/s | global throughput: 511.45 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:2:0] means: 0.01 sec | local throughput: 124.80 GiB/s | global throughput: 499.18 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
```

</details>

<details><summary>After (run 1)</summary>

```
[PRINT:2:0] means: 0.01 sec | local throughput: 132.01 GiB/s | global throughput: 528.06 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:1:0] means: 0.01 sec | local throughput: 127.69 GiB/s | global throughput: 510.74 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:0:0] means: 0.01 sec | local throughput: 129.64 GiB/s | global throughput: 518.55 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:3:0] means: 0.01 sec | local throughput: 128.10 GiB/s | global throughput: 512.41 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
```

</details>

As you may notice, there's some variability that makes it difficult to draw a line on whether performance is better or worse, but I think it's fair to say it has no regressions, the minimum throughput is generally higher, but maximum is lower. If we run multiple times we may see a different picture, in the results below we see both minimum and maximum throughput higher than before.

<details><summary>Before (run 2)</summary>

```
[PRINT:1:0] means: 0.01 sec | local throughput: 131.85 GiB/s | global throughput: 527.39 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:0:0] means: 0.01 sec | local throughput: 131.35 GiB/s | global throughput: 525.39 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:2:0] means: 0.01 sec | local throughput: 124.80 GiB/s | global throughput: 499.19 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:3:0] means: 0.01 sec | local throughput: 124.88 GiB/s | global throughput: 499.53 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
```

</details>

<details><summary>After (run 2)</summary>

```
[PRINT:1:0] means: 0.01 sec | local throughput: 129.12 GiB/s | global throughput: 516.48 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:0:0] means: 0.01 sec | local throughput: 129.59 GiB/s | global throughput: 518.36 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:2:0] means: 0.01 sec | local throughput: 134.02 GiB/s | global throughput: 536.08 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
[PRINT:3:0] means: 0.01 sec | local throughput: 129.40 GiB/s | global throughput: 517.61 GiB/s | in_parts: 1 | out_parts: 2 | nranks: 4
```

</details>

With that said, I propose we get this merged when we're happy with the way the code looks and tests pass and observe performance over the next few days in our nightly runs to ensure there were at least no regressions. WDYT?

madsbk commented on cpp/src/shuffler/shuffler.cpp:
```diff
@@ -480,6 +329,19 @@ Shuffler::Shuffler(
+                            size, br_->stream_pool().get_stream(), br_
+                        );
+                    },
+                    statistics
```
No, the order in which members are initialized in a constructor is guaranteed (it follows their declaration order in the class), but the evaluation order of the arguments passed to those members is not. Therefore, `std::move(statistics)` might be evaluated first.

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Fix control message

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Remove unused _op_id_ attribute

pentschev committed:
Prevent test_some call when using single comms

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Partial fix for op_id reuse

pentschev committed:
Support op_id reuse via MPE

pentschev committed:
Add polling mode UCXX progress to RapidsMPF's ProgressThread

pentschev committed:
Only send/expect receive termination from peers that communicated

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

pentschev committed:
Pre-distribute listener addresses during `barrier()`

pentschev committed:
Restructure control messages

pentschev committed:
Bulk-pack listener addresses into a single AM per rank

Replace the per-address ReplyListenerAddress messages with a single
DistributeListenerAddresses bulk message per destination rank during
barrier.  This reduces the number of AM sends from O(n^2) to O(n).

Also rename the remaining single-address message from
ReplyListenerAddress to RegisterListenerAddress, which better reflects
its purpose in the host/port init path.

Remove the now-unused endpoints_ insertion from the ranked
register_endpoint() overload (the ucp_ep_h-keyed map is no longer
read since the QueryListenerAddress removal).

pentschev committed:
Remove duplicate barrier() implementation

pentschev committed:
Add optional `data` arg to `Chunk::deserialize()`

pentschev reviewed (commented):
Thanks @wence- for the comments, I've addressed/responded them now, please let me know what you think and how you'd like to move forward. Btw, I've integrated changes from https://github.com/rapidsai/rapidsmpf/pull/934 here to unblock CI, but I think we would want to get that merged first, those changes are an overall improvement independent of MPE, but which we want for MPE.

pentschev committed:
Merge remote-tracking branch 'upstream/main' into shuffle-integrate-comms-interface

wence- commented on cpp/src/communicator/metadata_payload_exchange/tag.cpp:
```diff
@@ -119,64 +132,138 @@ TagMetadataPayloadExchange::recv() {
-        auto const [msg, src] = comm_->recv_any(metadata_tag_);
-        if (!msg) {
-            break;
+    // Use per-peer recv_from to avoid consuming messages belonging to a future
+    // collective on the same tag (see rapidsai/rapidsmpf#927).
+    for (Rank peer = 0; peer < nranks_; ++peer) {
+        if (peer == rank_) {
+            continue;
```
I think this is OK, i.e. to merge this, and then figure out the right interface to be slightly more general.

Just advertising the number of people we expect to receive from isn't sufficient because we still need to know _who_ we're going to receive from.

pentschev replied:
> Just advertising the number of people we expect to receive from isn't sufficient because we still need to know _who_ we're going to receive from.

If we know the number of peers we're going to receive from I think we don't need to know exactly who we should receive from, provided nobody is rogue and sends data we're not expecting. We can simply capture all messages addressed to the local worker based on the `op_id` (using a tag wildcard, or info the header), and I think in this way we don't actually _need_ to know who the peers are beforehand.

pentschev committed:
Merge branch 'main' into shuffle-integrate-comms-interface

madsbk commented on cpp/src/shuffler/shuffler.cpp:
```diff
@@ -25,17 +27,53 @@ namespace rapidsmpf::shuffler {
+
+    for (auto&& chunk : chunks) {
+        auto dst = peer_rank_fn(chunk);
+        auto metadata = *chunk.serialize();
```
I think we can avoid a copy?
```suggestion
        auto metadata = std::move(*chunk.serialize());
```

pentschev committed:
Prevent serialized metadata copying

Co-authored-by: Mads R. B. Kristensen <madsbk@gmail.com>

pentschev commented:
Thanks Lawrence and Mads for reviews and comments.

pentschev commented:
/merge
